### PR TITLE
Allow JSONConvertible to decode objects with required initialization parameters

### LINKED ISSUE #822: Change JSON deserialization to allow required parameters in model object constructors for models with the `JSONConvertible` mixin.
I came across this issue when tackling #817.

[This line](https://github.com/fastlane/ci/blob/master/app/shared/json_convertible.rb#L73) in `from_json!` in the `JSONConvertible` module creates a new object instance, with no parameters.

Hence, we can not have required parameters in our model constructors. Right now, we have [a bunch of optional parameters](https://github.com/fastlane/ci/blob/master/app/shared/models/user.rb#L14) with default values (usually set to `nil`).

Currently, in our deseralization method, we use `attr_accessor`s to set the values of the instance variables after object instantiation.

Some problems with this approch are:
- Some of these properties should be `attr_reader` only (such as `id`). Currently, a model's `id` could be arbitrarily set wherever since we're exposing a setter method.
- A model could be instantiated without the necessary parameters, and not fail explicitly (i.e., creating a `User` without an `email`, or without a `password_hash`).

screen shot 2018-05-07 at 1 25 37 pm

### PR DESCRIPTION
**Pull request by minuscorp**
Fixes #822

### CODE REVIEW

**Comment by taquitos** [comment]
Once this is merged, where do you think it would be helpful to make use of this?

**Comment by AndrewMcBurney** [comment]
Hey @taquitos, I described a good use case for it in https://github.com/fastlane/ci/issues/822#issue-320906594.

File: spec/shared/json_convertible_spec.rb [hunk 1]
```diff
@@ -77,6 +77,16 @@ def self.map_enumerable_type(enumerable_property_name: nil, current_json_object:
...
+
+  attr_reader :one_attribute
+
+  def initialize(one_attribute:)
+    @one_attribute = one_attribute
+  end
+end
+
 module FastlaneCI
   describe JSONConvertible do
```
Comment (taquitos) [inline spec/shared/json_convertible_spec.rb]: I think we need a few more specs to be safe:
1. flex the various failure cases (how many failure cases are there?)
2. objects that have a mix of required and non-required params
3. objects that have multiple required params
4. objects that have a mix of multiple required and multiple non-required params

Comment (minuscorp) [reply spec/shared/json_convertible_spec.rb]: Seems legit to me, will try to get this done for the weekend!

**Comment by minuscorp** [comment]
Ready to merge @taquitos ?

**Pull request merged**